File: frontend/src/shared/api/concept.ts
import { apiGet } from "./client";
import { mapBackendStockEnriched, type BackendStockEnriched } from "./stocks";
import type { StockRecord } from "@/shared/types";

export interface ConceptBoardItem {
  board_code: string;
  board_name: string;
  member_count: number;
  unresolved_count: number;
  priced_count: number;
  up_count: number;
  flat_count: number;
  down_count: number;
  avg_pct: number | null;
  main_net_inflow: number | null;
  main_net_ratio: number | null;
  lead_stock_name: string | null;
  lead_stock_code: string | null;
  lead_stock_pct: number | null;
  leaders: { symbol: string; name: string; change_percent: number | null }[];
}

export interface ConceptDetail {
  as_of: string;
  membership_as_of: string | null;
  source: string;
  degraded_reason: string | null;
  board: ConceptBoardItem;
  kpis: { zt_count: number; max_streak: number; leader_symbol: string | null; leader_name: string | null };
  echelons: Echelon[];
  unresolved_count: number;
  stock_count: number;
}

export interface SymbolConcept {
  board_code: string;
  board_name: string;
  pct_change: number | null;
}

export interface NewStockKpis {
  up_count: number;
  flat_count: number;
  down_count: number;
  unpriced_count: number;
  limit_up_count: number;
  unbroken_count: number;
  above_first_open_count: number;
  avg_pct: number | null;
}

export interface NewStockItem {
  symbol: string;
  name: string;
  exchange: string;
  list_date: string | null;
  listed_trade_days: number;
  pct_chg: number | null;
  close: number | null;
  turnover_rate: number | null;
  circ_mv: number | null;
  amount: number | null;
  streak: number | null;
  is_lu: boolean;
  never_broken: boolean | null;
  first_open: number | null;
  above_first_open: boolean | null;
}

export interface NewStocksResponse {
  as_of: string;
  membership_as_of: string | null;
  source: string;
  degraded_reason: string | null;
  /** `stock_id IS NULL` 的成分数（名录滞后披露）：既不在 items 也不进 KPI，>0 时卡片必须提示。 */
  unresolved_count: number;
  kpis: NewStockKpis;
  items: NewStockItem[];
}

/** 与后端 `app/schemas/limit_up.py:EchelonOut` 同形（板内梯队直接复用 `<LimitUpLadder>`）。 */
export interface EchelonStock {
  symbol: string;
  name: string;
  streak: number;
  days_span: number;
  boards_in_window: number;
  missing_days: number;
  amount?: number | null;
  seal_time?: string | null;
  seal_fund?: number | null;
  break_count?: number | null;
}

export interface Echelon {
  streak: number;
  label: string;
  stocks: EchelonStock[];
}

/**
 * `GET /api/v1/concepts` 的信封（与后端 `app/schemas/concept.py::ConceptListOut` 同键）。
 *
 * 与详情页不同，这里**不做 snake→camel 映射**：列表页的行就是 `ConceptBoardItem` 本身，
 * 多包一层只会让字段二次漂移。`as_of`（行情判据日）与 `membership_as_of`（成分快照日）是
 * 两个独立口径，列表页必须同时披露（历史口径声明是契约要求）。
 */
export interface ConceptListEnvelope {
  /** 行情（本地聚合）判据日；无行情时 null（配合 `degraded_reason`）。 */
  as_of: string | null;
  /** 全体成分行的 `max(last_seen_on)`：**成分**快照日，与 `as_of` 不是同一天。 */
  membership_as_of: string | null;
  price_source: "local_agg";
  /** 该 as_of 的东财资金流快照是否有行（数据集级）；无则 main_net_* 全为 null。 */
  flow_source: "em_clist" | null;
  /** 启用板块总数（与分页无关）。 */
  total: number;
  degraded_reason: string | null;
  items: ConceptBoardItem[];
}

/**
 * 概念板块列表（本地成分聚合，T-1/口径见 plans §2.2/§2.3）。
 *
 * 默认 `limit: 1000` 是刻意的：概念分类要「查看全部」（dev 库 504 板一次拉全），后端 `le=1000`
 * 与 `concept_service` 的全库聚合上限同量级。默认 `sort: "pct"` 保持 `avg_pct DESC NULLS LAST`。
 */
export function fetchConceptList(
  params: { sort?: "pct" | "inflow"; limit?: number; offset?: number } = {}
): Promise<ConceptListEnvelope> {
  const { sort = "pct", limit = 1000, offset = 0 } = params;
  return apiGet<ConceptListEnvelope>("/api/v1/concepts", { sort, limit, offset });
}

export function fetchConceptDetail(code: string) {
  return apiGet<ConceptDetail>(`/api/v1/concepts/${code}`);
}

export function fetchConceptStocks(code: string): Promise<StockRecord[]> {
  return apiGet<BackendStockEnriched[]>(`/api/v1/concepts/${code}/stocks`).then((r) =>
    r.map(mapBackendStockEnriched)
  );
}

export function fetchConceptsBySymbol(symbol: string) {
  return apiGet<{ as_of: string; membership_as_of: string | null; items: SymbolConcept[] }>(
    `/api/v1/concepts/by-symbol/${symbol}`
  );
}

export function fetchNewStocks(): Promise<NewStocksResponse> {
  return apiGet<NewStocksResponse>("/api/v1/new-stocks");
}

File: backend/app/schemas/concept.py
"""概念板块读接口响应模型（口径见 plans/2026-09-18-concept-boards-and-new-stocks.md §2.2/§2.3）。

两条字段级约定：

1. 资金流列（`main_net_inflow` / `main_net_ratio` / `lead_stock_*`）全部可空：东财快照一天只
   覆盖当日主力流入 Top-100 的震荡集，**缺失必须留 `None`（= `null`），绝不 0 填充**——0 会被
   前端读成"主力零净流入"而不是"没有这一行数据"。
2. `price_source` 与 `flow_source` 是两个独立字段（`local_agg` / `em_clist`）：价格来自本地
   成分聚合，资金流来自东财快照，两个源不允许混算成一个"综合"指标。
"""

from __future__ import annotations

from datetime import date
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.limit_up import EchelonOut


class BoardLeaderOut(BaseModel):
    """板块内涨幅前 2 成分（`pct_chg DESC NULLS LAST, symbol ASC` 的确定性取前二）。"""

    symbol: str
    name: str
    change_percent: float | None = None  # 无行情成分占位为 null（卡片侧会过滤掉）


class BoardItemOut(BaseModel):
    """§2.3 口径的板块行 + 东财资金流快照字段（快照缺行时全为 None）。"""

    board_code: str
    board_name: str
    member_count: int  # 本地 concept_members 行数（权威；不是采集时刻的东财 total）
    unresolved_count: int  # stock_id IS NULL 的成分数（名录滞后暴露面）
    priced_count: int  # 当日 pct_chg 非空的成分数（= up+flat+down）
    up_count: int
    flat_count: int
    down_count: int
    avg_pct: float | None = None  # 分母只有非空 pct_chg（n ≠ member_count），UI 注明
    main_net_inflow: float | None = None  # 元
    main_net_ratio: float | None = None  # %
    lead_stock_name: str | None = None
    lead_stock_code: str | None = None
    lead_stock_pct: float | None = None
    leaders: list[BoardLeaderOut] = Field(default_factory=list)


class ConceptListOut(BaseModel):
    """`GET /api/v1/concepts`（§2.2）：分页 items + 数据集级元信息。"""

    as_of: date | None  # 行情最新交易日（无行情 → None + degraded_reason="no_quotes"）
    # 全体成分行的 max(last_seen_on)：全集口径，不保证是本页这些板块的日期；无成分 → "no_members"
    membership_as_of: date | None
    price_source: Literal["local_agg"]
    flow_source: Literal["em_clist"] | None  # 该 as_of 的东财快照是否有行（数据集级，非本页命中数）
    total: int  # 启用板块总数（与分页无关）
    degraded_reason: str | None = None
    items: list[BoardItemOut] = Field(default_factory=list)


class ConceptKpisOut(BaseModel):
    """`GET /api/v1/concepts/{board_code}` 的板内连板 KPI（口径 = 梯队快照，见 §2.3）。

    `zt_count` / `max_streak` 是**同源**于 `/market/limit-up` 的快照：板内梯队被原样过滤，
    service 绝不重算 streak。龙头按 `(-streak, -amount, symbol)` 确定性裁决；无涨停成员时
    四个字段分别为 0/0/null/null（`max_streak=0` 是"无涨停"，不是缺失）。
    """

    zt_count: int
    max_streak: int
    leader_symbol: str | None = None
    leader_name: str | None = None


class ConceptDetailOut(BaseModel):
    """`GET /api/v1/concepts/{board_code}`（§2.2）：单板 `BoardItem` + 板内梯队 + KPI。

    `stock_count` / `unresolved_count` 与 `board.member_count` 同源（`concept_members` 行数）：
    `stock_count = board.member_count - unresolved_count`（= `stock_id IS NOT NULL` 的成分数）。
    """

    as_of: date | None
    membership_as_of: date | None  # **该板**的 max(last_seen_on)，不是全表值
    source: Literal["em_clist"]  # 成分名录来源（东财 clist），与资金流快照源无关
    degraded_reason: str | None = None  # "no_members" | "price_limits_missing" | None
    board: BoardItemOut
    kpis: ConceptKpisOut
    echelons: list[EchelonOut] = Field(default_factory=list)  # 板内过滤，形状与梯队卡一致
    unresolved_count: int
    stock_count: int


class ConceptBySymbolItemOut(BaseModel):
    """个股所属板块的一行：`pct_change` = 该板本地聚合 `avg_pct`（口径见 §2.3）。"""

    board_code: str
    board_name: str
    pct_change: float | None = None


class ConceptBySymbolOut(BaseModel):
    """`GET /api/v1/concepts/by-symbol/{symbol}`（§2.2）：未知 symbol 返回空 items，不是 404。"""

    as_of: date | None
    membership_as_of: date | None
    items: list[ConceptBySymbolItemOut] = Field(default_factory=list)
